orttraining/orttraining/test/training_api/trainer/trainer.cc

```diff
@@ -0,0 +1,360 @@
+
+#if defined(USE_CUDA) && defined(ENABLE_NVTX_PROFILE)
+        onnxruntime::profile::NvtxRangeCreator resetgrad_range(
+            "ResetGrad",
+            onnxruntime::profile::Color::Red);
+        resetgrad_range.Begin();
+#endif
+
+        ORT_RETURN_ON_ERROR(g_ort_training_api->ResetGrad(session));
+
+#if defined(USE_CUDA) && defined(ENABLE_NVTX_PROFILE)
+        resetgrad_range.End();
+#endif
+      }
+
+      if (do_eval && (batch_idx + 1) % params.eval_interval == 0) {
+        std::vector<OrtValue*> eval_results(eval_mode_output_count);
+        ORT_RETURN_ON_ERROR(g_ort_training_api->EvalStep(session, nullptr,
+                                                         inputs.size(), (const OrtValue* const*)inputs.data(),
+                                                         train_mode_output_count, eval_results.data()));
+      }
+
+      if ((batch_idx + 1) % params.checkpoint_interval == 0) {
+        // Save trained weights
+        PathString ckpt_file = ToPathString(params.output_dir + "/ckpt_" + params.model_name + std::to_string(batch_idx));
+        ORT_RETURN_ON_ERROR(g_ort_training_api->SaveCheckpoint(ckpt_file.c_str(), session, true));
+
+        // TODO(baiju): enable adding more properties to checkpoint
+        // state_to_save.property_bag.AddProperty<int64_t>(std::string("epoch"), epoch);
+      }
+      batch_idx++;
+
+      // release input ortvalues
+      for (size_t i = 0; i < inputs.size(); i++) {
+        g_ort_api->ReleaseValue(inputs[i]);
+      }
+
+      for (size_t i = 0; i < fetches.size(); i++) {
+        g_ort_api->ReleaseValue(fetches[i]);
+      }
+    }
+
+    data_loader.ResetIterateIndex();
+  }
+
+  // Save trained weights
+  PathString ckpt_file = ToPathString(params.output_dir + "/ckpt_" + params.model_name);
+  ORT_RETURN_ON_ERROR(g_ort_training_api->SaveCheckpoint(ckpt_file.c_str(), session, true));
+
+  auto end = std::chrono::high_resolution_clock::now();
+  std::chrono::duration<double> duration_seconds = end - end_to_end_start;
+  stabilized_total_end_to_end_time = duration_seconds.count();
+
+  std::cout << "Training completed - end to end latency: " << stabilized_total_end_to_end_time << "(s)" << std::endl;
+
+  // Delete all the ptrs
+  g_ort_training_api->ReleaseTrainingSession(session);
+
+#ifdef USE_CUDA
+  // Finally, don't forget to release the provider options
+  g_ort_api->ReleaseCUDAProviderOptions(cuda_options);
+#endif
+  g_ort_api->ReleaseSessionOptions(soptions);
+  g_ort_training_api->ReleaseCheckpointState(checkpoint_state);
+  g_ort_api->ReleaseEnv(env);
+
+  return 0;
+}
+
+int main(int argc, char* argv[]) {
+  // setup logger, be noted: LOGS_DEFAULT must be after logging manager initialization.
+  // This is to mitigate the issue " Attempt to use DefaultLogger but none has been registered".
+  // Need understand why the public CreateEnv did not get default logger ready.
+  Ort::Env env(ORT_LOGGING_LEVEL_WARNING, "e2e_test_runner");
```

Comment:
since there's only one Env (I think creating another one will return the same one), this doesn't need to be done in RunTraining as well